File: 251051/Birukov/JPA/src/main/java/com/jpa/service/NoteService.java
package com.jpa.service;

import com.jpa.dto.request.NoteRequestTo;
import com.jpa.dto.response.NoteResponseTo;
import com.jpa.entity.Note;
import com.jpa.entity.Tweet;
import com.jpa.mapper.NoteMapper;
import com.jpa.repository.NoteRepository;
import com.jpa.repository.TweetRepository;
import com.jpa.exception.NotFoundException;
import com.jpa.exception.ValidationException;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

@Service
@Transactional
@Validated
public class NoteService {
	
	@Autowired
	private final NoteRepository noteRepository;
	
	@Autowired
	private final TweetRepository tweetRepository;
	
	@Autowired
	private final NoteMapper noteMapper;
	
	public NoteService(NoteRepository noteRepository,
					   TweetRepository tweetRepository,
					   NoteMapper noteMapper) {
		this.noteRepository = noteRepository;
		this.tweetRepository = tweetRepository;
		this.noteMapper = noteMapper;
	}
	
	@Transactional
	public NoteResponseTo create(@Valid NoteRequestTo request) {
		validateNoteRequest(request);
		
		Tweet tweet = tweetRepository.findById(request.getTweetId())
			.orElseThrow(() -> new NotFoundException("Tweet not found: " + request.getTweetId()));
		
		Note note = noteMapper.toEntity(request);
		note.setTweet(tweet);
		Note saved = noteRepository.save(note);
		return noteMapper.toResponse(saved);
	}
	
	@Transactional
	public NoteResponseTo findById(Long id) {
		Note note = noteRepository.findById(id)
			.orElseThrow(() -> new NotFoundException("Note not found: " + id));
		return noteMapper.toResponse(note);
	}
	
	@Transactional
	public List<NoteResponseTo> findAll() {
		return noteRepository.findAll().stream()
			.map(noteMapper::toResponse)
			.toList();
	}
	
	@Transactional
	public NoteResponseTo update(Long id, @Valid NoteRequestTo request) {
		validateNoteRequest(request);
		
		Note existing = noteRepository.findById(id)
			.orElseThrow(() -> new NotFoundException("Note not found: " + id));
		
		Tweet tweet = tweetRepository.findById(request.getTweetId())
			.orElseThrow(() -> new NotFoundException("Tweet not found: " + request.getTweetId()));
		
		existing.setTweet(tweet);
		noteMapper.updateEntity(request, existing);
		Note updated = noteRepository.save(existing);
		return noteMapper.toResponse(updated);
	}
	
	@Transactional
	public void delete(Long id) {
		if (!noteRepository.existsById(id)) {
			throw new NotFoundException("Note not found: " + id);
		}
		noteRepository.deleteById(id);
	}
	
	@Transactional
	public List<NoteResponseTo> findByTweetId(Long TweetId) {
		if (!tweetRepository.existsById(TweetId)) {
			throw new NotFoundException("Tweet not found: " + TweetId);
		}
		
		return noteRepository.findByTweetId(TweetId).stream()
			.map(noteMapper::toResponse)
			.toList();
	}
	
	private void validateNoteRequest(NoteRequestTo request) {
		if (request.getTweetId() == null) {
			throw new ValidationException("Tweet ID is required");
		}
		if (request.getContent() == null || request.getContent().length() < 2 || request.getContent().length() > 2048) {
			throw new ValidationException("Content must be 2-2048 characters");
		}
	}
}

File: 351002/Abrashin/src/main/kotlin/com/example/distcomp/service/NoteService.kt
package com.example.distcomp.service

import com.example.distcomp.dto.request.NoteRequestTo
import com.example.distcomp.dto.response.NoteResponseTo
import com.example.distcomp.exception.NotFoundException
import com.example.distcomp.mapper.NoteMapper
import com.example.distcomp.repository.NoteRepository
import com.example.distcomp.repository.TweetRepository
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class NoteService(
    private val repository: NoteRepository,
    private val tweetRepository: TweetRepository,
    private val mapper: NoteMapper
) {
    @Transactional
    fun create(request: NoteRequestTo): NoteResponseTo {
        request.tweetId?.let {
            if (!tweetRepository.existsById(it)) throw NotFoundException("Tweet with id $it not found")
        }
        val entity = mapper.toEntity(request)
        entity.tweetId = request.tweetId
        val saved = repository.save(entity)
        return mapper.toResponse(saved)
    }

    fun getById(id: Long): NoteResponseTo {
        val entity = repository.findById(id) ?: throw NotFoundException("Note with id $id not found")
        return mapper.toResponse(entity)
    }

    fun getAll(page: Int, size: Int, sort: Array<String>): List<NoteResponseTo> {
        val sortOrder = if (sort.size >= 2) {
            Sort.by(Sort.Direction.fromString(sort[1]), sort[0])
        } else if (sort.isNotEmpty()) {
            Sort.by(sort[0])
        } else {
            Sort.unsorted()
        }
        val pageable = PageRequest.of(page, size, sortOrder)
        return repository.findAll(pageable).content.map { mapper.toResponse(it) }
    }

    @Transactional
    fun patch(id: Long, request: NoteRequestTo): NoteResponseTo {
        val existing = repository.findById(id) ?: throw NotFoundException("Note with id $id not found")
        
        request.tweetId?.let {
            if (!tweetRepository.existsById(it)) throw NotFoundException("Tweet with id $it not found")
            existing.tweetId = it
        }
        request.content?.let { existing.content = it }
        
        val saved = repository.save(existing)
        return mapper.toResponse(saved)
    }

    fun delete(id: Long) {
        if (!repository.deleteById(id)) {
            throw NotFoundException("Note with id $id not found")
        }
    }

    fun getNotesByTweetId(tweetId: Long): List<NoteResponseTo> {
        return repository.findByTweetId(tweetId).map { mapper.toResponse(it) }
    }
}